File: latest/LOL_sd/IC1.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.Arrays;
import java.util.Stack;
import java.util.StringTokenizer;
import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class IC1 {
	String[] keywords = { "auto", "break", "case", "char", "const", "continue",
			"default", "do", "double", "else", "enum", "extern", "float",
			"for", "goto", "if", "int", "include", "long", "main", "register",
			"return", "short", "signed", "sizeof", "static", "struct",
			"switch", "typedef", "union", "unsigned", "void", "volatile",
			"while" };
	String[] ariOp = { "+", "-", "*", "/" };
	String[] relOp = { "!", ">", "==", "<", "!=", "<=", ">=" };

	FileReader fr = null;
	BufferedReader br = null;

	FileReader fr1;
	BufferedReader br1;

	static String value = null;
	int j = 0, length = 0, k = 0, l = 0, stacktop = 0, index = 0;
	String parTable[][] = new String[36][22];

	String productions[] = { "", "S -> A S", "S -> I S", "S -> A", "S -> I",
			"A -> id = E ;", "I -> while ( C ) { S }", "C -> id relop id",
			"E -> E + T", "E -> E - T", "E -> T", "T -> T * F", "T -> T / F",
			"T -> F", "F -> ( E )", "F -> id" };
	String[] ic_reduce = { "na", "na", "na", "na", "na", "=", "na", "r", "+",
			"-", "na", "*", "/", "na", "na", "na" };
	String terNonter[] = { "id", ";", "=", "{", "}", "(", ")", "relop",
			"while", "+", "-", "*", "/", "$", "S", "A", "I", "C", "E", "T", "F" };

	static int readOneLine = 0;
	int goInLoop = 0, readLastChar = 0,cunt=0;
	Stack st = new Stack();
	Stack symbol = new Stack();
	static Stack inputString = new Stack();
	static Stack valueStack = new Stack();

	public static void main(String[] args) throws IOException {

		String str = null, strng = null, str1 = null, line = null, line1 = null, splitLine[] = null;
		int j = 0, k = 0;

		IC1 sy = new IC1();

		// reading input file
		FileReader fr = new FileReader("Input1.txt");
		BufferedReader br = new BufferedReader(fr);
		FileReader fr2 = new FileReader("Input1.txt");
		BufferedReader br2 = new BufferedReader(fr2);

		// reading parsing table
		FileReader fr1 = new FileReader("Parse.csv");
		BufferedReader br1 = new BufferedReader(fr1);
		line = br1.readLine();
		// loading parse table in an array
		j = 0;
		while ((line = br1.readLine()) != null) {
			splitLine = line.split(",");
			k = 0;
			for (int i = 1; i < splitLine.length; i++) {
				sy.parTable[j][k] = splitLine[i];
				k++;
			}
			j++;
		}

		for (int g = 0; g < sy.parTable.length; g++) {
			//System.out.print(g + "\t");
			for (int h = 0; h < splitLine.length; h++) {
			//	System.out.print(sy.parTable[g][h] + "\t");
			}
			//System.out.println();
		}
		System.out.println("\nCode Gen\n");
		sy.st.push("0");
		inputString.push("$");
		sy.value = "started";

//		System.out.print("\n\nStack\tSymbol\tInput String\tAction\n");
//		System.out.print(Arrays.toString(sy.st.toArray()) + "\t");
//		System.out.print(Arrays.toString(sy.symbol.toArray()) + "\t");
//		System.out.print(Arrays.toString(inputString.toArray()) + "\t");

		// execute token wise
		strng = br2.readLine();
		// str = br.readLine();
		while ((str = br.readLine()) != null) {
				if ((br2.readLine() == null)) {
				readOneLine = 1;
			}
			String token = sy.lex(str);
		}
		if (value.equals("ACCEPT")) {
			System.out.println("STRING ACCEPTED");
		} else {
			System.out.println("STRING REJECTED");
		}
	}

	public String lex(String str) throws IOException {
//		System.out.println("in lex function");
		String str1 = null, array[];
		if (str.trim().length() > 0) {
			// remove extra spaces
			StringTokenizer st = new StringTokenizer(str, " ");
			StringBuffer sb = new StringBuffer();
			while (st.hasMoreElements()) {
				sb.append(st.nextElement()).append(" ");
			}
			str = sb.toString().trim();
			// remove tabs or new line
			str = str.replaceAll("\t", "");
			str1 = str.replaceAll("\n", "");

			// check for single line comment
			if (str1.contains("//")) {
//				System.out.println("\n" + str1);
//				System.out.println("\nSingle line comment");
				int ind = str1.indexOf("//");
				str1 = str1.substring(0, ind);
//				System.out.println("New String " + str1);
			}

			// check for multiple line comment
			if (str1.contains("/*")) {
//				System.out.println("\nMultiple line comment");
//				System.out.println(str1);
				str1 = str1.substring(str1.indexOf("/*"), str1.length());
				while (!str1.contains("*/")) {

					str1 = br.readLine();
//					System.out.println(str1);
					str1.replaceAll(str1, "");
				}
				int a = str1.indexOf("*/");
				str1 = str1.substring(a + 2, str1.length());
			}

			// split with spaces
			array = str1.split(" ");
//			System.out.println("\n" + str1 + "\n");
			// readOneLine=0;
			for (int i = 0; i < array.length; i++) {
				if ((i == (array.length - 1)) && (readOneLine == 1)) {
					readLastChar = 1;
				}
				if (checkKeyword(array[i])) {
//					System.out.println("Keyword : " + array[i]);
					syntax(array[i], array[i]);
				} else if (checkAriOperator(array[i])) {
//					System.out.println("Arithmatic operator : " + array[i]);
					syntax(array[i], array[i]);
				} else if (checkRelOp(array[i])) {
//					System.out.println("Relational operator : " + array[i]);
					syntax("relop", array[i]);
				} else if (array[i].equals("=")) {
//					System.out.println("Assignment Operator : " + array[i]);
					syntax(array[i], array[i]);
				} else if ((array[i].equals("{")) || (array[i].equals("}"))
						|| (array[i].equals("(")) || (array[i].equals(")"))) {
//					System.out.println("Brackets : " + array[i]);
					syntax(array[i], array[i]);
				} else if ((array[i].equals("||")) || (array[i].equals("&&"))) {
//					System.out.println("Logical Operators : " + array[i]);
					syntax(array[i], array[i]);
				} else if ((array[i].equals(";")) || (array[i].equals(","))) {
//					System.out.println("Delimitor : " + array[i]);
					syntax(array[i], array[i]);
				} else if (!Character.isDigit(array[i].charAt(0))) {
//					System.out.println("Identifier : " + array[i]);
					syntax("id", array[i]);
				} else if (checkNumber(array[i])) {
//					System.out.println("Integer Number : " + array[i]);
					syntax(array[i], array[i]);
				} else if (checkNumber1(array[i])) {
//					System.out.println("Float Number : " + array[i]);
					syntax(array[i], array[i]);
				} else {
//					System.out.println("Invalid identifier!!!!! : " + array[i]);
				}
			}
//			System.out.println("Read one line  and readOneLine = "
//					+ readOneLine);
			readOneLine = 1;
		}
		return (null);
	}

	public void syntax(String token, String token1) {
		String value1=null,value2=null,value3=null;
		inputString.push(token);
		System.out.print(Arrays.toString(inputString.toArray()));
		int i = 0;
		while (((!inputString.peek().equals("$")) || (goInLoop == 1) || (readLastChar == 1)
				&& (!((value.equals("ACCEPT")) || (value.equals("null")))))
				&& i < 100) {
			String input = (String) inputString.peek();
			for (k = 0; k < terNonter.length; k++) {
				if (input.equals(terNonter[k])) {
					index = k;
					break;
				}
			}

			String top = (String) st.peek();
			stacktop = Integer.parseInt(top);

			// get value from table
			value = parTable[stacktop][index];
			// System.out.println(" value is    " + value + "  stacktop  " +
			// stacktop
			// + "  input is " + input);

			if (!((value.equals("ACCEPT")) || (value.equals("null")))) {
				if (value.startsWith("S")) {
//					System.out.print("\nShift operation -> read " + value);
					// System.out.println("\n");
					if (value.length() == 2)
						value = value.substring(1, 2);
					else
						value = value.substring(1, 3);

					st.push(value);
					symbol.push(input);
					inputString.pop();
					if (token == "id" || token == "relop") {
						valueStack.push(token1);
					}

				} else if (value.startsWith("R")) {
					// System.out.print("Reduce operation -> read " + value);

					if (value.length() == 2)
						value = value.substring(1, 2);
					else
						value = value.substring(1, 3);

					int num = Integer.parseInt(value);

					String reductn = ic_reduce[num];
					
					System.out.println(Arrays.toString(valueStack.toArray()) + "\t");
					switch (reductn) {
					case "na":
						break;
					case "+":
								value2 = "" + valueStack.pop();
								value1 = "" + valueStack.pop();
								
								System.out.println("\nt" + (cunt++) + "= " + value1+ " + " + value2 + "\n");
								valueStack.push("t" + (cunt - 1));
								System.out.print(Arrays.toString(valueStack.toArray()) + "\t");
								break;
					case "-":
						value2 = "" + valueStack.pop();
						value1 = "" + valueStack.pop();
						System.out.println("\nt" + (cunt++) + "= " + value1+ " - " + value2 + "\n");
						valueStack.push("t" + (cunt - 1));
						System.out.print( Arrays.toString(valueStack.toArray()) + "\t");
						break;
					case "*":
						value2 = "" + valueStack.pop();
						value1 = "" + valueStack.pop();
						System.out.println("\nt" + (cunt++) + "= " + value1+ " * " + value2 + "\n");
						valueStack.push("t" + (cunt - 1));
						System.out.print(Arrays.toString(valueStack.toArray()) + "\t");
						break;
					case "/":
						value2 = "" + valueStack.pop();
						value1 = "" + valueStack.pop();
						System.out.println("\nt" + (cunt++) + "= " + value1+ " / " + value2 + "\n");
						valueStack.push("t" + (cunt - 1));
						System.out.print(Arrays.toString(valueStack.toArray()) + "\t");
						break;
					case "=":
						value2 = "" + valueStack.pop();
						value1 = "" + valueStack.pop();
						System.out.println("\n"+value1+ " = " + value2 + "\n");
						//valueStack.push("t" + (cunt - 1));
						System.out.print(Arrays.toString(valueStack.toArray()) + "\t");
						valueStack.empty();
						break;
					case "relop":
						value3 = "" + valueStack.pop();
						value2 = "" + valueStack.pop();
						value1 = "" + valueStack.pop();
						System.out.println("\nt" + (cunt++) + "= " + value1 + value2 + value3 + "\n");
                        valueStack.push("t"+(cunt));       
                        System.out.print(Arrays.toString(valueStack.toArray()) + "\t");
                        break;
					}

					String prod = productions[num];
					String rhs = prod.substring(5, prod.length());
					String[] rhsArr = rhs.split(" ");
					prod = prod.substring(0, 1);
					// System.out
					// .println("  " + symbol.peek() + " reduced to " + prod);
					// System.out.println("\n");
					for (int o = 0; o < rhsArr.length; o++) {
//						System.out.println(rhsArr[o]);
						st.pop();
					}
					symbol.pop();
					symbol.push(prod);

					// checking after reduce and non terminal
					// String symtop = (String) symbol.peek();
					for (k = 0; k < terNonter.length; k++) {
						if (prod.equals(terNonter[k])) {
							index = k;
							break;
						}
					}
					top = (String) st.peek();
					stacktop = Integer.parseInt(top);
					value = parTable[stacktop][index];
					st.push(value);
					// System.out.println("\n\nafter reduce\n");
					// System.out.println("stck : "+stacktop);
					// System.out.println("input : "+ input);
					// System.out.println("symbol : "+symbol.peek()
					// +"uska index "+index);
					// System.out.println("value[stack][input]  "+value
					// +"  that is pushed to stack");
				} else {
					st.push(value);
					symbol.push(input);
					inputString.pop();
				}

//				System.out.print("\n\n" + Arrays.toString(st.toArray()) + "\t");
//				System.out.print(Arrays.toString(symbol.toArray()) + "\t");
//				System.out.print(Arrays.toString(inputString.toArray()) + "\t");
//				System.out.println("\n");
			}
			if ((inputString.peek().equals("$")) && (readOneLine == 1)) {
				goInLoop = 1;
			}
			i++;

		}

	}

	public boolean checkNumber(String str) {
		try {
			int e = Integer.parseInt(str);

		} catch (NumberFormatException e) {
			return false;
		}
		return true;
	}

	public boolean checkNumber1(String str) {
		try {
			float w = Float.parseFloat(str);
		} catch (NumberFormatException e) {
			return false;

		}
		return true;
	}

	public boolean checkKeyword(String str) {
		boolean res = false;
		for (int k = 0; k < keywords.length; k++) {
			if (str.equals(keywords[k])) {
				res = true;
				break;
			}
		}
		return res;
	}

	public boolean checkAriOperator(String str) {
		boolean res = false;
		for (int k = 0; k < ariOp.length; k++) {
			if (str.equals(ariOp[k])) {
				res = true;
				break;
			}
		}
		return res;
	}

	public boolean checkRelOp(String str) {
		boolean res = false;
		for (int k = 0; k < relOp.length; k++) {
			if (str.equals(relOp[k])) {
				res = true;
				break;
			}
		}
		return res;
	}

}

File: ICGeneration/ICGeneration/IC1.cs
using System;
using System.Collections;
using System.Text;
using System.IO;

public class IC1
{
	internal string[] keywords = new string[] {"auto", "break", "case", "char", "const", "continue", "default", "do", "double", "else", "enum", "extern", "float", "for", "goto", "if", "int", "include", "long", "main", "register", "return", "short", "signed", "sizeof", "static", "struct", "switch", "typedef", "union", "unsigned", "void", "volatile", "while"};
	internal string[] ariOp = new string[] {"+", "-", "*", "/"};
	internal string[] relOp = new string[] {"!", ">", "==", "<", "!=", "<=", ">="};

	internal StreamReader fr = null;
	internal StreamReader br = null;

	internal StreamReader fr1;
	internal StreamReader br1;

	internal static string value = null;

	internal int j = 0, length = 0, k = 0, l = 0, stacktop = 0, index = 0;
//JAVA TO C# CONVERTER NOTE: The following call to the 'RectangularArrays' helper class reproduces the rectangular array initialization that is automatic in Java:
//ORIGINAL LINE: internal string[][] parTable = new string[36][22];
	internal string[][] parTable = RectangularArrays.ReturnRectangularStringArray(36, 22);

	internal string[] productions = {};
	internal string[] ic_reduce = new string[] {"na", "na", "na", "na", "na", "=", "na", "r", "+", "-", "na", "*", "/", "na", "na", "na"};
	internal string[] terNonter = {};

	internal static int readOneLine = 0;
	internal int goInLoop = 0, readLastChar = 0, cunt = 0;
	internal Stack st = new Stack();
	internal Stack symbol = new Stack();
	internal static Stack inputString = new Stack();
	internal static Stack valueStack = new Stack();

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in .NET:
//ORIGINAL LINE: public static void main(String[] args) throws java.io.IOException
	public static void Main(string[] args)
	{

		string str = null; string strng = null; string str1 = null; string line = null; string line1 = null; string[] splitLine = null;
		int j = 0, k = 0;

		IC1 sy = new IC1();

		// reading input file
		StreamReader fr = new StreamReader("Input1.txt");
		//StreamReader br = new StreamReader(fr);
		StreamReader fr2 = new StreamReader("Input1.txt");
		//StreamReader br2 = new StreamReader(fr2);

		// reading parsing table
		StreamReader fr1 = new StreamReader("Parse.csv");
		//StreamReader br1 = new StreamReader(fr1);
		line = fr1.ReadLine();
		// loading parse table in an array
		j = 0;
		while ((line = fr1.ReadLine()) != null)
		{
			splitLine = line.Split(",", true);
			k = 0;
			for (int i = 1; i < splitLine.Length; i++)
			{
				sy.parTable[j][k] = splitLine[i];
				k++;
			}
			j++;
		}

		for (int g = 0; g < sy.parTable.Length; g++)
		{
			Console.Write(g + "\t");
			for (int h = 0; h < splitLine.Length; h++)
			{
				Console.Write(sy.parTable[g][h] + "\t");
			}
			Console.WriteLine();
		}
		Console.WriteLine("\nCode Gen\n");
		sy.st.Push("0");
		inputString.Push("$");
		IC1.value = "started";

//		System.out.print("\n\nStack\tSymbol\tInput String\tAction\n");
//		System.out.print(Arrays.toString(sy.st.toArray()) + "\t");
//		System.out.print(Arrays.toString(sy.symbol.toArray()) + "\t");
//		System.out.print(Arrays.toString(inputString.toArray()) + "\t");

		// execute token wise
		strng = fr2.ReadLine();
		// str = br.ReadLine();
		while ((str = fr.ReadLine()) != null)
		{
				if ((fr2.ReadLine() == null))
				{
				readOneLine = 1;
				}
			string token = sy.lex(str);
		}
		if (value.Equals("ACCEPT"))
		{
			Console.WriteLine("STRING ACCEPTED");
		}
		else
		{
			Console.WriteLine("STRING REJECTED");
		}
	}

//JAVA TO C# CONVERTER WARNING: Method 'throws' clauses are not available in .NET:
//ORIGINAL LINE: public String lex(String str) throws java.io.IOException
	public virtual string lex(string str)
	{
//		System.out.println("in lex function");
		string str1 = null; string[] array ;
		if (str.Trim().Length > 0)
		{
			// remove extra spaces
			//StringTokenizer st = new StringTokenizer(str, " ");
			StringBuilder sb = new StringBuilder();
			//while (st.hasMoreElements())
			//{
			//	sb.Append(st.nextElement()).Append(" ");
			//}
			//str = sb.ToString().Trim();
			// remove tabs or new line
			//str = str.replaceAll("\t", "");
			//str1 = str.replaceAll("\n", "");

			var tokens = str.Split (' ');
			foreach (var tok in tokens) {
				sb.Append (tok + " ");
				str = sb.ToString ().Trim ();
				str = str.Replace ("\t", "");
				str = str.Replace ("\r", "");
				str = str.Replace ("\n", "");
				str1 = str;
			}
			// check for single line comment
			if (str1.Contains("//"))
			{
//				System.out.println("\n" + str1);
//				System.out.println("\nSingle line comment");
				int ind = str1.IndexOf("//", StringComparison.Ordinal);
				str1 = str1.Substring(0, ind);
//				System.out.println("New String " + str1);
			}

			// check for multiple line comment
			if (str1.Contains("/*"))
			{
//				System.out.println("\nMultiple line comment");
//				System.out.println(str1);
				str1 = StringHelperClass.SubstringSpecial(str1, str1.IndexOf("/*", StringComparison.Ordinal), str1.Length);
				while (!str1.Contains("*/"))
				{

					str1 = br.ReadLine();
//					System.out.println(str1);
					str1.Replace(str1, "");
				}
				int a = str1.IndexOf("*/", StringComparison.Ordinal);
				str1 = StringHelperClass.SubstringSpecial(str1, a + 2, str1.Length);
			}

			// split with spaces
			array = str1.Split(" ", true);
//			System.out.println("\n" + str1 + "\n");
			// readOneLine=0;
			for (int i = 0; i < array.Length; i++)
			{
				if ((i == (array.Length - 1)) && (readOneLine == 1))
				{
					readLastChar = 1;
				}
				if (checkKeyword(array[i]))
				{
//					System.out.println("Keyword : " + array[i]);
					syntax(array[i], array[i]);
				}
				else if (checkAriOperator(array[i]))
				{
//					System.out.println("Arithmatic operator : " + array[i]);
					syntax(array[i], array[i]);
				}
				else if (checkRelOp(array[i]))
				{
//					System.out.println("Relational operator : " + array[i]);
					syntax("relop", array[i]);
				}
				else if (array[i].Equals("="))
				{
//					System.out.println("Assignment Operator : " + array[i]);
					syntax(array[i], array[i]);
				}
				else if ((array[i].Equals("{")) || (array[i].Equals("}")) || (array[i].Equals("(")) || (array[i].Equals(")")))
				{
//					System.out.println("Brackets : " + array[i]);
					syntax(array[i], array[i]);
				}
				else if ((array[i].Equals("||")) || (array[i].Equals("&&")))
				{
//					System.out.println("Logical Operators : " + array[i]);
					syntax(array[i], array[i]);
				}
				else if ((array[i].Equals(";")) || (array[i].Equals(",")))
				{
//					System.out.println("Delimitor : " + array[i]);
					syntax(array[i], array[i]);
				}
				else if (!char.IsDigit(array[i][0]))
				{
//					System.out.println("Identifier : " + array[i]);
					syntax("id", array[i]);
				}
				else if (checkNumber(array[i]))
				{
//					System.out.println("Integer Number : " + array[i]);
					syntax(array[i], array[i]);
				}
				else if (checkNumber1(array[i]))
				{
//					System.out.println("Float Number : " + array[i]);
					syntax(array[i], array[i]);
				}
				else
				{
//					System.out.println("Invalid identifier!!!!! : " + array[i]);
				}
			}
//			System.out.println("Read one line  and readOneLine = "
//					+ readOneLine);
			readOneLine = 1;
		}
		return (null);
	}

	public virtual void syntax(string token, string token1)
	{
		string value1 = null, value2 = null, value3 = null;
		inputString.Push(token);
//		System.out.print(Arrays.toString(inputString.toArray()));
		int i = 0;
		while (((!inputString.Peek().Equals("$")) || (goInLoop == 1) || (readLastChar == 1) && (!((value.Equals("ACCEPT")) || (value.Equals("null"))))) && i < 100)
		{
			string input = (string) inputString.Peek();
			for (k = 0; k < terNonter.Length; k++)
			{
				if (input.Equals(terNonter[k]))
				{
					index = k;
					break;
				}
			}

			string top = (string) st.Peek();
			stacktop = int.Parse(top);

			// get value from table
			value = parTable[stacktop][index];
			// System.out.println(" value is    " + value + "  stacktop  " +
			// stacktop
			// + "  input is " + input);

			if (!((value.Equals("ACCEPT")) || (value.Equals("null"))))
			{
				if (value.StartsWith("S", StringComparison.Ordinal))
				{
//					System.out.print("\nShift operation -> read " + value);
					// System.out.println("\n");
					if (value.Length == 2)
					{
						value = value.Substring(1, 1);
					}
					else
					{
						value = value.Substring(1, 2);
					}

					st.Push(value);
					symbol.Push(input);
					inputString.Pop();
					if (token == "id" || token == "relop")
					{
						valueStack.Push(token1);
					}

				}
				else if (value.StartsWith("R", StringComparison.Ordinal))
				{
					// System.out.print("Reduce operation -> read " + value);

					if (value.Length == 2)
					{
						value = value.Substring(1, 1);
					}
					else
					{
						value = value.Substring(1, 2);
					}

					int num = int.Parse(value);

					string reductn = ic_reduce[num];

					//Console.WriteLine(Arrays.ToString(valueStack.ToArray()) + "\t");
					switch (reductn)
					{
					case "na":
						break;
					case "+":
								value2 = "" + valueStack.Pop();
								value1 = "" + valueStack.Pop();

								Console.WriteLine("\nt" + (cunt++) + "= " + value1 + " + " + value2 + "\n");
								valueStack.Push("t" + (cunt - 1));
						//		Console.Write(Arrays.ToString(valueStack.ToArray()) + "\t");
								break;
					case "-":
						value2 = "" + valueStack.Pop();
						value1 = "" + valueStack.Pop();
						Console.WriteLine("\nt" + (cunt++) + "= " + value1 + " - " + value2 + "\n");
						valueStack.Push("t" + (cunt - 1));
						//Console.Write(Arrays.ToString(valueStack.ToArray()) + "\t");
						break;
					case "*":
						value2 = "" + valueStack.Pop();
						value1 = "" + valueStack.Pop();
						Console.WriteLine("\nt" + (cunt++) + "= " + value1 + " * " + value2 + "\n");
						valueStack.Push("t" + (cunt - 1));
						//Console.Write(Arrays.ToString(valueStack.ToArray()) + "\t");
						break;
					case "/":
						value2 = "" + valueStack.Pop();
						value1 = "" + valueStack.Pop();
						Console.WriteLine("\nt" + (cunt++) + "= " + value1 + " / " + value2 + "\n");
						valueStack.Push("t" + (cunt - 1));
						//Console.Write(Arrays.ToString(valueStack.ToArray()) + "\t");
						break;
					case "=":
						value2 = "" + valueStack.Pop();
						value1 = "" + valueStack.Pop();
						Console.WriteLine("\n" + value1 + " = " + value2 + "\n");
						//valueStack.push("t" + (cunt - 1));
						//Console.Write(Arrays.ToString(valueStack.ToArray()) + "\t");
						//valueStack.Count == 0;
						break;
					case "relop":
						value3 = "" + valueStack.Pop();
						value2 = "" + valueStack.Pop();
						value1 = "" + valueStack.Pop();
						Console.WriteLine("\nt" + (cunt++) + "= " + value1 + value2 + value3 + "\n");
						valueStack.Push("t" + (cunt));
						//Console.Write(Arrays.ToString(valueStack.ToArray()) + "\t");
						break;
					}

					string prod = productions[num];
					string rhs = prod.Substring(5, prod.Length - 5);
					string[] rhsArr = rhs.Split(" ", true);
					prod = prod.Substring(0, 1);
					// System.out
					// .println("  " + symbol.peek() + " reduced to " + prod);
					// System.out.println("\n");
					for (int o = 0; o < rhsArr.Length; o++)
					{
//						System.out.println(rhsArr[o]);
						st.Pop();
					}
					symbol.Pop();
					symbol.Push(prod);

					// checking after reduce and non terminal
					// String symtop = (String) symbol.peek();
					for (k = 0; k < terNonter.Length; k++)
					{
						if (prod.Equals(terNonter[k]))
						{
							index = k;
							break;
						}
					}
					top = (string) st.Peek();
					stacktop = int.Parse(top);
					value = parTable[stacktop][index];
					st.Push(value);
					// System.out.println("\n\nafter reduce\n");
					// System.out.println("stck : "+stacktop);
					// System.out.println("input : "+ input);
					// System.out.println("symbol : "+symbol.peek()
					// +"uska index "+index);
					// System.out.println("value[stack][input]  "+value
					// +"  that is pushed to stack");
				}
				else
				{
					st.Push(value);
					symbol.Push(input);
					inputString.Pop();
				}

//				System.out.print("\n\n" + Arrays.toString(st.toArray()) + "\t");
//				System.out.print(Arrays.toString(symbol.toArray()) + "\t");
//				System.out.print(Arrays.toString(inputString.toArray()) + "\t");
//				System.out.println("\n");
			}
			if ((inputString.Peek().Equals("$")) && (readOneLine == 1))
			{
				goInLoop = 1;
			}
			i++;

		}

	}

	public virtual bool checkNumber(string str)
	{
		try
		{
			int e = int.Parse(str);

		}
		catch (System.FormatException)
		{
			return false;
		}
		return true;
	}

	public virtual bool checkNumber1(string str)
	{
		try
		{
			float w = float.Parse(str);
		}
		catch (System.FormatException)
		{
			return false;

		}
		return true;
	}

	public virtual bool checkKeyword(string str)
	{
		bool res = false;
		for (int k = 0; k < keywords.Length; k++)
		{
			if (str.Equals(keywords[k]))
			{
				res = true;
				break;
			}
		}
		return res;
	}

	public virtual bool checkAriOperator(string str)
	{
		bool res = false;
		for (int k = 0; k < ariOp.Length; k++)
		{
			if (str.Equals(ariOp[k]))
			{
				res = true;
				break;
			}
		}
		return res;
	}

	public virtual bool checkRelOp(string str)
	{
		bool res = false;
		for (int k = 0; k < relOp.Length; k++)
		{
			if (str.Equals(relOp[k]))
			{
				res = true;
				break;
			}
		}
		return res;
	}

}
